python/mxnet/optimizer/optimizer.py

```diff
@@ -1259,53 +1260,101 @@ def __init__(self, learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-6,
         self.lower_bound = lower_bound
         self.upper_bound = upper_bound
         self.bias_correction = bias_correction
-
+        self.aggregate_num = max(1, min(45, int(os.getenv('MXNET_OPTIMIZER_AGGREGATION_SIZE', "45"))))
 
     def create_state(self, index, weight):
         stype = weight.stype
-        return (zeros(weight.shape, weight.context, dtype=numpy.float32, stype=stype),
-                zeros(weight.shape, weight.context, dtype=numpy.float32, stype=stype))
+        dtype = weight.dtype
+        return (zeros(weight.shape, weight.context, dtype=dtype, stype=stype),
+                zeros(weight.shape, weight.context, dtype=dtype, stype=stype))
 
     def _update_impl(self, index, weight, grad, state, multi_precision=False):
-        assert(isinstance(weight, NDArray))
-        assert(isinstance(grad, NDArray))
-        self._update_count(index)
-        lr = self._get_lr(index)
-        wd = self._get_wd(index)
-        t = self._index_update_count[index]
-
         kwargs = {'beta1': self.beta1, 'beta2': self.beta2, 'epsilon': self.epsilon,
-                  'bias_correction': self.bias_correction, 't': t,
+                  'bias_correction': self.bias_correction,
                   'rescale_grad': self.rescale_grad}
 
-        if self.clip_gradient:
-            kwargs['clip_gradient'] = self.clip_gradient
-
-        if multi_precision:
-            mean, var = state[1]
-            weight32 = state[0]
-            g = mp_lamb_update_phase1(weight, grad, mean, var, weight32, wd=wd, **kwargs)
+        if self.aggregate_num <= 1 or not isinstance(index, (tuple, list)):
```

Comment:
MXNET_OPTIMIZER_AGGREGATION_SIZE=1 does not seem to work because weight will be a list instead of NDArray